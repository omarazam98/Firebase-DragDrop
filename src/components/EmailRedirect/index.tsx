import React from 'react';
import { withAPI } from '@winwin/api-firebase';

export const EmailRedirect = (props) => {
  const handleClick = () => props.history.push('/login');
  const handleSendClick = () => {
    const redirect = window['_env_'].NODE_ENV === 'production' ? 'http://www.winwinhomesharing.com'
      : window['_env_'].REACT_APP_EMAIL_REDIRECT || 'http://integration.scoutmastersforever.com';
    props.api.auth.currentUser().sendEmailVerification({ url: redirect });
  };

  return (
    <div>
      Email has been sent to verify account. <br/>
      Already verified?
      <button onClick={handleClick} id="redirect"> Log In</button> <br/>
      Didn't receive the email?
      <button onClick={handleSendClick} id="resend"> Send Again</button>
    </div>
  );
};
export default withAPI(EmailRedirect);
