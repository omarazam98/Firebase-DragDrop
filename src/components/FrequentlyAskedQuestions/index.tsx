import React, { useEffect, useState } from 'react';
import { withAPI } from '@winwin/api-firebase';
import FrequentQuestions from './FrequentQuestions';

interface FrequentlyAskedQuestionsPageProps {
  api: any;
}

const FrequentlyAskedQuestionsPage: React.FC<FrequentlyAskedQuestionsPageProps> = (props) => {
  const [questionList, setQuestionList] = useState([]);

  useEffect(() => {
    const frequentQuestionsCollection = props.api.data.faq.getAll();
    frequentQuestionsCollection.then(questions =>
      setQuestionList(questionList => questions));
  },        [props.api.data.faq]);

  const handleSubmit = function (question:string, email?:string) {
    props.api.data.faq.addQuestion(question, email);
  };

  return (
    <div>
      <FrequentQuestions questionList={questionList} handleSubmit={handleSubmit}/>
    </div>
  );

};

export default withAPI(FrequentlyAskedQuestionsPage);
