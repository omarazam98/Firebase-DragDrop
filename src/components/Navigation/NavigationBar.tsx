import React from 'react';
import { withRouter, Link } from 'react-router-dom';
import { MenuList, MenuItem } from '@material-ui/core';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
    },
    paper: {
      marginRight: theme.spacing(2),
    },
    content: {
      paddingLeft: theme.spacing(6),
    },
  }),
);
interface NavigationBarProps {
  location: any;
  routesList: any;
}

export const NavigationBar: React.FC<NavigationBarProps> = (props) => {
  useStyles();
  return (
    <MenuList>
      {props.routesList.map((route: any, key: number) => {
        return (
          <MenuItem component={Link} to={route.path}
                    selected={props.location.pathname === route.path}
                    key={key}>
            {route.name}<br/>
          </MenuItem>
        );
      })}
    </MenuList>
  );
};

export default withRouter(NavigationBar);
