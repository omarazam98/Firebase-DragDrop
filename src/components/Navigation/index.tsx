// Main routing component.
// If a new page is added, be sure to add the route to src/constants/routes
import React from 'react';
import { BrowserRouter as Router, Route, Link } from 'react-router-dom';
import { ROUTES, NAVBAR_ROUTES, RouteType } from '../../constants/routes';
import SplitPane from 'react-split-pane';
import { withAPI } from '@winwin/api-firebase';

interface NavState {
  loggedIn: boolean;
  emailVerified: boolean;
  isMounted: boolean;
}
interface NavProps {
  api: any;
}
export class Navigation extends React.Component<NavProps, NavState> {

  constructor(props) {
    super(props);
    this.state = {
      loggedIn: false,
      emailVerified: false,
      isMounted: false,
    };
    this.links = this.links.bind(this);
  }

  componentDidMount() {
    this.setState(() => {
      return {
        isMounted: true,
      };
    });
    this.props.api.auth.onAuthStateChanged((user) => {
      if (this.state.isMounted) {
        this.setState(() => {
          return {
            loggedIn: user ? true : false,
            emailVerified: (user && user.emailVerified) ? true : false,
          };
        });
      }
    });
  }

  componentWillUnmount() {
    this.setState(() => {
      return {
        isMounted: false,
      };
    });
  }

  links() {
    // can't push JSX on an empty array, so init with a div
    const allLinks = [<div/>];
    NAVBAR_ROUTES.forEach((route: any) =>
      ((this.state.loggedIn && this.state.emailVerified) || !route.authRequired) && allLinks.push(
      <Link to={route.path}>{route.name}<br/></Link>));
    return allLinks;
  }

  render() {
    return (
      <div>
        <Router>
          <SplitPane split="vertical" minSize={50} defaultSize={100}>
            {/* Link creates the object that a user can click on to go to another page */}
            <div>
              {this.links()}
              {this.state.loggedIn && <button id="signOut" onClick={() => {
                this.props.api.auth.signOut();
              }}>Log Out</button> }
            </div>
            {/* Route indicates what component should be shown, based on what is linked */}
            {/* Map routes from links to their components */}
            <div>
              {ROUTES.map((route: RouteType, index) => (
                <Route
                  key={index}
                  path={route.path}
                  exact={route.exact}
                  component={route.component}
                />
              ))}
            </div>
          </SplitPane>
        </Router>
      </div>
    );
  }
}

export default withAPI(Navigation);
