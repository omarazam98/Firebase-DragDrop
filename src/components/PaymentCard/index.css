.credit-card {
    display: flex;
    flex-direction: column;
    justify-content: space-between;

    width: 50vw;
    height: 32vw;
    padding: 25px;
    border-radius: 15px;
    max-width: 400px;
    max-height: 250px;
    color: white;

    background-image: linear-gradient(25deg, #0f509e, #1399cd);
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.5);
}

.credit-card__number {
    display: flex;
    justify-content: center;

    font-family: 'Arial', sans-serif;
    font-size: 2em;
}

.credit-card__info {
    display: flex;
    justify-content: space-between;

    font-family: 'Arial', sans-serif;
    font-size: 2em;
}

.credit-card__info_label {
    font-size: 0.4em;
}

.credit-card__info_expiry {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
}
