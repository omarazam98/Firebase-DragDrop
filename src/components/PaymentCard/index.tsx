import React from 'react';
import './index.css';
import { withAPI } from '@winwin/api-firebase';

interface PaymentMethods {
  type: string;
  number: number;
}

interface UserType {
  displayName: string;
  uid: string;
}
interface PaymentProps {
  api: any;
}
interface PaymentState {
  paymentMethod: PaymentMethods | undefined;
  user: UserType | undefined;
}
const INITIAL_STATE = {
  paymentMethod: undefined,
  user: undefined,
};

export class PaymentCard extends React.Component<PaymentProps, PaymentState>{
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  componentDidMount() {
    const user = this.props.api.auth.currentUser();
    if (user) {
      this.props.api.data.users.getPaymentMethods(user.uid).then((payments) => {
        this.setState(() => {
          return {
            user,
            paymentMethod: payments,
          };
        });
      });
    }
  }

  render() {
    return(
      <div className="credit-card">
        <div className="credit-card__logo">Win-Win Payment Method</div>
        <div className="credit-card__number">
          {this.state.paymentMethod && `XXXX XXXX XXXX ${this.state.paymentMethod.number}`}
        </div>

        <div className="credit-card__info">
          <div className="credit-card__info_name">
            <div className="credit-card__info_label">CARDHOLDER'S NAME</div>
            <div>{this.state.user && this.state.user.displayName}</div>
          </div>

          <div className="credit-card__info_expiry">
            <button>Change</button>
          </div>
        </div>

      </div>
    );
  }
}

export default withAPI(PaymentCard);
