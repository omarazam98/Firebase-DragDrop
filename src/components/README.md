## What are components?
Components are reusable, independent pieces of user interface.

They are a core concept in react, so if this is news to you check out the docs at:
https://reactjs.org/docs/components-and-props.html 

## How do we use components?
Components can be completely independent vanilla react components. In that case data is passed to them by their parent with props, and is modified locally or by their children using this.state. These are generally called presentational components, and have no concept of redux.

However if components to interact with data from either Redux, or the database they do so in two ways.

1. Receiving data from redux through mapStateToProps
2. Dispatching action to redux through mapDispatchToProps

Components that either need to receive data from redux or dispatch actions to redux need to be connected to the redux store. This is done as follows.

**If any of this is confusing this resource is quite good https://blog.logrocket.com/react-redux-connect-when-and-how-to-use-it-f2a1edab2013/**

For this example assume that this component has been called as a decendant of <Provider> 
We have a redux store that has this structure describing a garage:

  {
    cars: [
      {},
      {},
      {}
    ],
    open: false
  }

And an action creator that looks like this

  const toggleDoor = () => {
    return {
      type: 'TOGGLE_DOOR',
    }
  }

The component would then look like this

  import React, { Component } from 'react';
  import { connect } from 'react-redux';
  import { changeName } from '../actions/fileThatHasThisCreatorInIt'

  // This class should be exported as a named export to be used in testing.
  // However as explained at the end of this example this export will not be the one used in source code
  export class SomeComponent extends Component {
    constructor(props){
      super(props)

      // if you are unfamiliar with binding check this out https://reactjs.org/docs/handling-events.html
      this.openCloseGarage = this.openCloseGarage.bind(this);
    }

    openCloseGarage() {
      // toggleGarageDoor is available as a prop, because we have used mapDispatchToProps seen below
      this.props.toggleGarageDoor();
    }

    render() {
      return (
        <div>
          <button onClick={this.openCloseGarage}>
            // garageOpen is available as a prop beause we provided it through mapStateToProps seen below
            {this.props.garageOpen}
          </button>
        </div>
      )
    }

    // This is where the redux specific details are added

    // mapStateToProps is used to subscribe to data from the redux store.
    // It allows data from the store to be accessed as a prop in the component
    // The first variable is the redux store state, and the second are the props that are provided by the component constructor (Normal props not mapped to redux state)

    const mapStateToProps = (reduxState, currentProps) => {
      return {
        // this is where the prop garageOpen is mapped to the 'open' field of theh state of the redux store
        garageOpen: state.open,
      }
    }

    // mapDispatchToProps is used to gain access to dispatching actions as a component prop
    // The first variable is the store.dispatch command, and the second are the props that are providd by the component constructor (Normal props not mapped to redux state)

    const mapDispatchToProps = (dispatch, currentProps) => {
      return {
        // this is where the prop toggleGarageDoor is mapped to the dispatching of the action toggleDoor
        // For this one ensure that you are declaring an inline function that calls dispatch with the action creator as the argument
        toggleGarageDoor: () => { dispatch(toggleDoor()) }
      }
    }

    // Finally the functions that were defined above need to be connected to the component.
    // Ensure that this connected component is what is exported as default and used in the source code.
    // However it is useful to export the unconnected component as a named export for testing

    connect(mapStateToProps, mapDispatchToProps)(SomeComponent)

  }