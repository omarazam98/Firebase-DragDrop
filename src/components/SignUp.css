#phoneNumber {
    background-color: orange
}

#email {
    background-color: red

}

#passwordOne {
    background-color: green;

}

#passwordTwo {
    background-color: blue

}

#phoneNumber p.ws-errormessage::after {
    content:"Please enter the phone number in the form (xxx) xxx-xxxx";
}

.error { border: 1px solid red; }

.form-control:valid~.invalid-feedback {display: none;}
