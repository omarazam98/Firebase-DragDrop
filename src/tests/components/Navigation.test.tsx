import { Navigation } from '../../components/Navigation';
import * as React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme/build';

const options = {
  lifecycleExperimental: false,
  disableLifecycleMethods: true,
};

const mockAPI = {
  auth: {
    signOut: jest.fn(),
  },
};

test('Render the navigation', () => {
  const test = <Navigation />;
  const wrapper = shallow(test, options);
  expect(wrapper).toMatchSnapshot();
});

test('Sign out button calls sign out', () => {
  const test = <Navigation api={ mockAPI } />;
  const wrapper = shallow(test, options);
  wrapper.find('#signOut').simulate('click');
  expect(mockAPI.auth.signOut).toHaveBeenCalled();
});
