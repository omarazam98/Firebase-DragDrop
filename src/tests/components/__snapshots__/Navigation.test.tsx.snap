// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the navigation authed 1`] = `
<div>
  <BrowserRouter>
    <SplitPane
      allowResize={true}
      defaultSize={100}
      minSize={50}
      pane1ClassName=""
      pane2ClassName=""
      paneClassName=""
      primary="first"
      split="vertical"
    >
      <div>
        <div />
        <Link
          to="/signup"
        >
          Sign Up
          <br />
        </Link>
        <Link
          to="/login"
        >
          Log In
          <br />
        </Link>
        <Link
          to="/dashboard"
        >
          Dashboard
          <br />
        </Link>
        <Link
          to="/payment"
        >
          My Wallet
          <br />
        </Link>
        <button
          id="signOut"
          onClick={[Function]}
        >
          Log Out
        </button>
      </div>
      <div>
        <Route
          component={[Function]}
          exact={true}
          key="0"
          path="/signup"
        />
        <Route
          component={[Function]}
          exact={true}
          key="1"
          path="/login"
        />
        <Route
          component={[Function]}
          exact={true}
          key="2"
          path="/dashboard"
        />
        <Route
          component={[Function]}
          exact={true}
          key="3"
          path="/payment"
        />
        <Route
          component={[Function]}
          exact={true}
          key="4"
          path="/email"
        />
      </div>
    </SplitPane>
  </BrowserRouter>
</div>
`;

exports[`Render the navigation unauthed 1`] = `
<div>
  <BrowserRouter>
    <SplitPane
      allowResize={true}
      defaultSize={100}
      minSize={50}
      pane1ClassName=""
      pane2ClassName=""
      paneClassName=""
      primary="first"
      split="vertical"
    >
      <div>
        <div />
        <Link
          to="/signup"
        >
          Sign Up
          <br />
        </Link>
        <Link
          to="/login"
        >
          Log In
          <br />
        </Link>
      </div>
      <div>
        <Route
          component={[Function]}
          exact={true}
          key="0"
          path="/signup"
        />
        <Route
          component={[Function]}
          exact={true}
          key="1"
          path="/login"
        />
        <Route
          component={[Function]}
          exact={true}
          key="2"
          path="/dashboard"
        />
        <Route
          component={[Function]}
          exact={true}
          key="3"
          path="/email"
        />
      </div>
    </SplitPane>
  </BrowserRouter>
</div>
`;
